### ISSUE TITLE
Mixed modeling of Chinese and English, there are many errors in post-inference recognition of onnx model.

### ISSUE THREAD

**Issue by kellkwang** [k2-fsa/icefall#1925]
Using the 8k non-streaming model trained with 8k data, the pt model inference is normal.
tokens.txt , the training script and export onnx script are as follows. There are many repeated words in the inference results.

/home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/c-api/c-api.cc:GetOfflineRecognizerConfig:497 OfflineRecognizerConfig(feat_config=FeatureExtractorConfig(sampling_rate=8000, feature_dim=80, low_freq=20, high_freq=-400, dither=0), model_config=OfflineModelConfig(transducer=OfflineTransducerModelConfig(encoder_filename="/data/test_zipformer/model/backvalley_cce/encoder-epoch-10-avg-1.onnx", decoder_filename="/data/test_zipformer/model/backvalley_cce/decoder-epoch-10-avg-1.onnx", joiner_filename="/data/test_zipformer/model/backvalley_cce/joiner-epoch-10-avg-1.onnx"), paraformer=OfflineParaformerModelConfig(model=""), nemo_ctc=OfflineNemoEncDecCtcModelConfig(model=""), whisper=OfflineWhisperModelConfig(encoder="", decoder="", language="", task="transcribe", tail_paddings=-1), fire_red_asr=OfflineFireRedAsrModelConfig(encoder="", decoder=""), tdnn=OfflineTdnnModelConfig(model=""), zipformer_ctc=OfflineZipformerCtcModelConfig(model=""), wenet_ctc=OfflineWenetCtcModelConfig(model=""), sense_voice=OfflineSenseVoiceModelConfig(model="", language="", use_itn=False), moonshine=OfflineMoonshineModelConfig(preprocessor="", encoder="", uncached_decoder="", cached_decoder=""), telespeech_ctc="", tokens="/data/test_zipformer/model/backvalley_cce/tokens.txt", num_threads=1, debug=True, provider="cpu", model_type="", modeling_unit="cjkchar", bpe_vocab=""), lm_config=OfflineLMConfig(model="", scale=1), ctc_fst_decoder_config=OfflineCtcFstDecoderConfig(graph="", max_active=3000), decoding_method="greedy_search", max_active_paths=4, hotwords_file="", hotwords_score=1.5, blank_penalty=0, rule_fsts="", rule_fars="")

/home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/csrc/offline-transducer-model.cc:InitEncoder:176 ---encoder---
model_author=k2-fsa
model_type=zipformer2
comment=non-streaming zipformer2
version=1

/home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/csrc/offline-transducer-model.cc:InitDecoder:197 ---decoder---
vocab_size=6031
context_size=2

/home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/csrc/offline-transducer-model.cc:InitJoiner:221 ---joiner---
joiner_dim=512

A2_0	是阳阳春烟烟景景景大文文底底底四四四的林林更是绿鲜鲜鲜秀秀秀媚媚诗诗盎盎
Decode start : 2025-04-21 19:44:39.970
Decode end : 2025-04-21 19:44:40.695
Total Wave duration 9812(ms)
Total Decode duration : 725(ms)
AVG Decode RTF : 0.073889

cat tokens.txt
<blk> 0
<sos/eos> 1
<unk> 2
四 3
个 4
人 5
▁O 6
K 7
哎 8
你 9
应 10
该 11
...
果 105
▁VI 106
D 107
E 108
O 109
▁GA 110
ME 111
S 112
...
意 156
思 157
▁I 158
▁DON 159
' 160
T 161
▁THINK 162
▁THAT 163
▁IS 164
▁RIGHT 165
▁TWO 166
▁F 167
ET 168
▁HIGH 169
高 170
除 171
了 172
哪 173
▁NOW 174
...
茼 6029
蛉 6030
#0 6031
#1 6032

cat train_cce.sh
#!/bin/bash

export CUDA_VISIBLE_DEVICES="0,1,2,3,4,5,6,7"

./zipformer/train_load.py \
  --world-size 8 \
  --num-epochs 80 \
  --start-epoch 10 \
  --use-fp16 0 \
  --context-size 1 \
  --exp-dir exp/zipformer_cce \
  --lang-dir data/lang_char_ce \
  --enable-musan 0 \
  --base-lr 0.045 \
  --lr-batches 7500 \
  --lr-epochs 18 \
  --causal 0 \
  --spec-aug-time-warp-factor 20 \
  --num-encoder-layers 2,2,4,4,4,2 \
  --feedforward-dim 768,1024,1536,1536,1536,768 \
  --encoder-dim 256,384,512,512,512,256 \
  --encoder-unmasked-dim 192,192,256,256,256,192 \
  --master-port 12356 \
  --max-duration 350

cat  export_onnx_cce.sh 
#!/bin/bash

./zipformer/export-onnx.py \
  --tokens data/lang_char_ce/tokens.txt \
  --use-averaged-model 0 \
  --epoch 10 \
  --avg 1 \
  --exp-dir exp/zipformer_cce \
  --causal 0 \
  --num-encoder-layers 2,2,4,4,4,2 \
  --feedforward-dim 768,1024,1536,1536,1536,768 \
  --encoder-dim 256,384,512,512,512,256 \
  --encoder-unmasked-dim 192,192,256,256,256,192

(zipformer) kell@ubuntu:~/backvalley/icefall/egs/aishell/ASR$ ./export_onnx_cce.sh 
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/scaling.py:1306: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  def forward(ctx, x, y, alpha):
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/scaling.py:1315: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  def backward(ctx, ans_grad):
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/scaling.py:1512: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.
  def forward(
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/scaling.py:1551: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.
  def backward(ctx, ans_grad: Tensor):
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
2025-04-21 16:45:00,869 INFO [export-onnx.py:442] device: cuda:0
2025-04-21 16:45:00,885 INFO [export-onnx.py:448] {
  "attention_decoder_attention_dim": 512,
  "attention_decoder_dim": 512,
  "attention_decoder_feedforward_dim": 2048,
  "attention_decoder_num_heads": 8,
  "attention_decoder_num_layers": 6,
  "avg": 1,
  "batch_idx_train": 0,
  "best_train_epoch": -1,
  "best_train_loss": Infinity,
  "best_valid_epoch": -1,
  "best_valid_loss": Infinity,
  "blank_id": 0,
  "causal": false,
  "chunk_size": "16,32,64,-1",
  "cnn_module_kernel": "31,31,15,15,15,31",
  "context_size": 2,
  "decoder_dim": 512,
  "downsampling_factor": "1,2,4,8,4,2",
  "encoder_dim": "256,384,512,512,512,256",
  "encoder_unmasked_dim": "192,192,256,256,256,192",
  "env_info": {
    "IP address": "100.95.129.58",
    "hostname": "ubuntu",
    "icefall-git-branch": null,
    "icefall-git-date": null,
    "icefall-git-sha1": null,
    "icefall-path": "/home/kell/anaconda3/envs/zipformer/lib/python3.9/site-packages/icefall-1.0-py3.9.egg",
    "k2-build-type": "Release",
    "k2-git-date": "Tue Oct 29 09:02:19 2024",
    "k2-git-sha1": "75e2ed6b2fd87c22b7f3f34bad48a69984bb8755",
    "k2-path": "/home/kell/anaconda3/envs/zipformer/lib/python3.9/site-packages/k2/__init__.py",
    "k2-version": "1.24.4",
    "k2-with-cuda": true,
    "lhotse-path": "/home/kell/anaconda3/envs/zipformer/lib/python3.9/site-packages/lhotse/__init__.py",
    "lhotse-version": "1.30.0.dev+git.13c7616.clean",
    "python-version": "3.9",
    "torch-cuda-available": true,
    "torch-cuda-version": "12.1",
    "torch-version": "2.4.0+cu121"
  },
  "epoch": 10,
  "exp_dir": "exp/zipformer_cce",
  "feature_dim": 80,
  "feedforward_dim": "768,1024,1536,1536,1536,768",
  "fp16": false,
  "ignore_id": -1,
  "iter": 0,
  "joiner_dim": 512,
  "label_smoothing": 0.1,
  "left_context_frames": "64,128,256,-1",
  "log_interval": 50,
  "num_encoder_layers": "2,2,4,4,4,2",
  "num_heads": "4,4,4,8,4,4",
  "pos_dim": 48,
  "pos_head_dim": "4",
  "query_head_dim": "32",
  "reset_interval": 200,
  "subsampling_factor": 4,
  "tokens": "data/lang_char_ce/tokens.txt",
  "use_attention_decoder": false,
  "use_averaged_model": false,
  "use_cr_ctc": false,
  "use_ctc": false,
  "use_transducer": true,
  "valid_interval": 3000,
  "value_head_dim": "12",
  "vocab_size": 6031,
  "warm_step": 2000
}
2025-04-21 16:45:00,886 INFO [export-onnx.py:450] About to create model
2025-04-21 16:45:03,961 INFO [checkpoint.py:112] Loading checkpoint from exp/zipformer_cce/epoch-10.pt
/home/kell/anaconda3/envs/zipformer/lib/python3.9/site-packages/icefall-1.0-py3.9.egg/icefall/checkpoint.py:113: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(filename, map_location="cpu")
2025-04-21 16:45:05,989 INFO [export-onnx.py:554] encoder parameters: 105602297
2025-04-21 16:45:05,989 INFO [export-onnx.py:555] decoder parameters: 3354624
2025-04-21 16:45:05,989 INFO [export-onnx.py:556] joiner parameters: 3093903
2025-04-21 16:45:05,989 INFO [export-onnx.py:557] total parameters: 112050824
2025-04-21 16:45:05,989 INFO [export-onnx.py:568] Exporting encoder
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/scaling.py:1487: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  zero = torch.tensor(0.0, dtype=x.dtype, device=x.device)
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/scaling.py:1413: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  zero = torch.tensor(0.0, dtype=x.dtype, device=x.device)
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/scaling.py:473: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert x.shape[self.channel_dim] == self.num_channels
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/subsampling.py:330: TracerWarning: Converting a tensor to a Python number might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert x.size(1) == x_lens.max().item(), (x.size(1), x_lens.max())
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/subsampling.py:330: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert x.size(1) == x_lens.max().item(), (x.size(1), x_lens.max())
/home/kell/anaconda3/envs/zipformer/lib/python3.9/site-packages/icefall-1.0-py3.9.egg/icefall/utils.py:1370: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  max_len = max(max_len, lengths.max())
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/scaling.py:1658: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if num_channels <= x.shape[-1]:
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/zipformer.py:1633: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert p.shape[-1] == num_heads * pos_head_dim, (
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/zipformer.py:1717: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert attn_scores.shape == (num_heads, batch_size, seq_len, seq_len)
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/zipformer.py:1728: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert key_padding_mask.shape == (
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/zipformer.py:2145: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert attn_weights.shape == (num_heads, batch_size, seq_len, seq_len)
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/zipformer.py:1935: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert attn_weights.shape == (num_heads, batch_size, seq_len, seq_len)
/home/kell/backvalley/icefall/egs/librispeech/ASR/zipformer/zipformer.py:1359: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert src.shape[0] == d_seq_len * ds, (src.shape, d_seq_len, ds)
/home/kell/anaconda3/envs/zipformer/lib/python3.9/site-packages/torch/onnx/utils.py:884: UserWarning: no signature found for <torch.ScriptMethod object at 0x7ffa3abdd270>, skipping _decide_input_format
  warnings.warn(f"{e}, skipping _decide_input_format")
2025-04-21 16:45:56,890 INFO [export-onnx.py:332] meta_data: {'model_type': 'zipformer2', 'version': '1', 'model_author': 'k2-fsa', 'comment': 'non-streaming zipformer2'}
2025-04-21 16:45:58,865 INFO [export-onnx.py:575] Exported encoder to exp/zipformer_cce/encoder-epoch-10-avg-1.onnx
2025-04-21 16:45:58,866 INFO [export-onnx.py:577] Exporting decoder
2025-04-21 16:45:59,085 INFO [export-onnx.py:584] Exported decoder to exp/zipformer_cce/decoder-epoch-10-avg-1.onnx
2025-04-21 16:45:59,085 INFO [export-onnx.py:586] Exporting joiner
2025-04-21 16:45:59,085 INFO [export-onnx.py:402] joiner dim: 512
2025-04-21 16:45:59,235 INFO [export-onnx.py:593] Exported joiner to exp/zipformer_cce/joiner-epoch-10-avg-1.onnx
2025-04-21 16:45:59,235 INFO [export-onnx.py:616] Generate int8 quantization models
2025-04-21 16:46:03,196 WARNING [quantize.py:557] Please consider to run pre-processing before quantization. Refer to example: 

... // Omit quantified logs

2025-04-21 16:46:24,710 INFO [onnx_quantizer.py:591] Quantization parameters for tensor:"/Tanh_output_0" not specified

**Comment by csukuangfj** [comment]
Here is part of your training command
```
/zipformer/train_load.py
--world-size 8
--num-epochs 80
--start-epoch 10
--use-fp16 0
--context-size 1
```

Here is part of your export log
```
"cnn_module_kernel": "31,31,15,15,15,31",
"context_size": 2,
```

You can see `context_size` is different betwen training and exporting.

Please never use default arguments.

Please re-export your model with `--context-size 1`.

**Comment by kellkwang** [comment]
> Here is part of your training command
> 
> ```
> /zipformer/train_load.py
> --world-size 8
> --num-epochs 80
> --start-epoch 10
> --use-fp16 0
> --context-size 1
> ```
> 
> Here is part of your export log
> 
> ```
> "cnn_module_kernel": "31,31,15,15,15,31",
> "context_size": 2,
> ```
> 
> You can see `context_size` is different betwen training and exporting.
> 
> Please never use default arguments.
> 
> Please re-export your model with `--context-size 1`.

yes, that's my fault.
after modification --context-size 1, the exported model is normal.
